legend title example at XYPlot component

moontrip opened the pull request:
examined and tested to add legend title using XYPlot component. If acceptable, then I will add this approach to other plot components. A screenshot with an exemplary title of `legend title example` is attached:

![legend-title-example](https://user-images.githubusercontent.com/12802305/121989081-8154a380-cd69-11eb-9a0b-0804e27c6dbc.png)

bobular commented on src/plots/XYPlot.tsx:
```diff
@@ -69,9 +73,10 @@ export default function XYPlot(props: ScatterplotProps) {
   } = props;
 
-  const layout: Partial<Layout> = {
+  const layout: Partial<Layout> & layoutLegend = {
```
I think it would be good to stick with our unspoken rule of capitalising type names - and also rename it `LayoutLegendTitle`

bobular commented on src/types/general.ts:
```diff
@@ -40,3 +43,19 @@ export type TimeDelta = { value: number; unit: string };
+      side?: 'top' | 'left' | 'top left';
+    };
+  };
+}
```
I have two suggestions to tidy this up a bit, by removing all the question marks, and wrapping each inner level in `Partial<>` like they do in the index.d.ts of plotly.js

Then line 79 of XYPlot.tsx can be 
```
  const layout: Partial<Layout> & Partial<LayoutLegendTitle> = {
```

However, another solution is to define all the props as required, then use RecursivePartial defined [here on stackoverflow](https://stackoverflow.com/a/51365037/1909761) 
as follows:
```
  const layout: Partial<Layout> & RecursivePartial<LayoutLegendTitle> = {
```

bobular replied:
Reading more of https://stackoverflow.com/questions/41980195/recursive-partialt-in-typescript before I closed the tab... it seems like that is a risky path to take.

Probably best to keep the '?'s or use `Partial<>` like here in plotly's index.d.ts
```
export interface Layout {
    colorway: string[];
    title:
        | string
        | Partial<{
              text: string;
              font: Partial<Font>;
              xref: 'container' | 'paper';
              yref: 'container' | 'paper';
              x: number;
              y: number;
              xanchor: 'auto' | 'left' | 'center' | 'right';
              yanchor: 'auto' | 'top' | 'middle' | 'bottom';
              pad: Partial<Padding>;
          }>;
    titlefont: Partial<Font>;
    ...
```

bobular commented on src/types/general.ts:
```diff
@@ -40,3 +43,19 @@ export type TimeDelta = { value: number; unit: string };
+
+// extend legend type: somehow plotly.js did not define legend.title type
+// see https://plotly.com/javascript/reference/layout/#layout-legend-title
+export interface layoutLegend extends Partial<Legend> {
```
With my testing, the `extends Partial<Legend>` was not necessary.

bobular reviewed (changes requested):
Thanks DK. That's a good workaround to the annoying lapse in plotly TS.  Should we report a bug?

My minimal requested change is the renaming of the type with a capital letter. The other TS gurus may want to comment on the style suggestions regarding partial/optional props.

bobular commented:
Also, maybe move the extra type definition to a new *.ts file?   `src/types/plotly-omissions.ts` maybe?

Then if/when the problem is solved upstream, we can just remove the file, rather than risk (very small risk, I know) a merge conflict in general.ts.

moontrip commented:
@bobular thank you for your review and suggestions. As discussed in Slack, I will wait for others' comment and if these are good, then you may commit your local changes to this PR: certainly I can do that too if you want :)

bobular commented:
> @bobular thank you for your review and suggestions. As discussed in Slack, I will wait for others' comment and if these are good, then you may commit your local changes to this PR: certainly I can do that too if you want :)

Sorry I've already rolled them back - I was just playing around to see what would compile.  But I sent you the diff in case there's anything useful in there.

moontrip commented:
@bobular yes you already described them at our slack chat: anyway your explanations are clear so I can handle it anyway :) As you suggested in the end, will make another ts file to have the legend.title type and then use the original(?) one with ? for now.

moontrip commented:
@bobular I have updated this PR based on your suggestions such as changing the prop name to `LayoutLegendTitle`, placing its  type definition to the new file, `plotly-omissions.ts`, and removed `Legend` type import. As mentioned in my previous comment, I left the use of `?` for now following your comments.

bobular reviewed (approved):
Great.  I will probably also use plotly-omissions.ts in the general refactoring. It seems that type definitions are missing in several places.

moontrip commented:
Thank you for your reviews @jtlong3rd and @bobular 👍 

@bobular As I said in the last meeting, I will wait for your works (not merge) even if this ticket is approved :)

bobular commented:
> Thank you for your reviews @jtlong3rd and @bobular +1
> 
> @bobular As I said in the last meeting, I will wait for your works (not merge) even if this ticket is approved :)

Thanks for the consideration @moontrip 

I have just checked this PR against my proposed changes and they will not cause any conflicts, so please merge to `main` and then I can merge from here if needed.  Thanks!

moontrip commented:
@bobular thank you for your check for conflicts :) 
@dmfalke any review/comment ?

dmfalke commented:
@moontrip, sorry for the delayed review. I read through the comments, and I think this looks good. One thing I'd like to consider is if it is possible to use relative font sizes for the legend title (and possibly other text that plotly draws). It's not urgent, but in the future we might want/need to be able to scale a plot, etc.

moontrip commented:
@dmfalke Thank you for your review and comment 👍 Actually I tried to use relative font size like `em` in there, however, the `legend.title.font.size` parameter at plot does not accept string: only allows number. The legend title is actually handled as `SVG.text` in the plotly: if I remember correctly, `SVG.text` has font-size attribute, thus I think it can also be handled via CSS later.

bobular commented:
I agree. During the refactor tidy-up I'm removing as much
arbitrary/absolute styling as possible in order to pave the way for
something more flexible/maintainable.

On Thu, Jun 17, 2021 at 3:15 PM DK ***@***.***> wrote:

> @dmfalke <https://github.com/dmfalke> Thank you for your review and
> comment 👍 Actually I tried to use relative font size like em in there,
> however, the legend.title.font.size parameter at plot does not accept
> string: only allows number. The legend title is actually handled as
> SVG.text in the plotly: if I remember correctly, SVG.text has font-size
> attribute, thus I think it can also be handled via CSS later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/VEuPathDB/web-components/pull/152#issuecomment-863276989>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACLLH4D6V2ORRY6F2STGCDTTH7O5ANCNFSM46WMS3RA>
> .
>
